src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+
+  if (nulls.empty()) {
+    return std::make_shared<ExpressionResult<R>>(std::move(values));
+  } else {
+    return std::make_shared<ExpressionResult<R>>(std::move(values), std::move(nulls));
+  }
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_select_expression(
+    const PQPSelectExpression& select_expression) {
+  const auto select_result_tables = _evaluate_select_expression_to_tables(select_expression);
+  const auto select_result_columns = _prune_tables_to_expression_results<R>(select_result_tables);
+
+  std::vector<R> result_values(select_result_columns.size());
+
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < select_result_columns.size(); ++chunk_offset) {
+    Assert(select_result_columns[chunk_offset]->size() == 1,
+           "Expected precisely one to be returned from SelectExpression");
+    result_values[chunk_offset] = select_result_columns[chunk_offset]->value(0);
+  }
+
+  if (select_expression.is_nullable()) {
+    std::vector<bool> result_nulls(select_result_columns.size());
+
+    for (auto chunk_offset = ChunkOffset{0}; chunk_offset < select_result_columns.size(); ++chunk_offset) {
+      result_nulls[chunk_offset] = select_result_columns[chunk_offset]->null(0);
+    }
+    return std::make_shared<ExpressionResult<R>>(std::move(result_values), std::move(result_nulls));
+  } else {
+    return std::make_shared<ExpressionResult<R>>(std::move(result_values));
+  }
+}
+
+std::vector<std::shared_ptr<const Table>> ExpressionEvaluator::_evaluate_select_expression_to_tables(
+    const PQPSelectExpression& expression) {
+  // If the SelectExpression is uncorrelated, evaluating it once is sufficient
+  if (expression.parameters.empty()) {
+    return {_evaluate_select_expression_for_row(expression, ChunkOffset{0})};
+  }
+
+  // Make sure all Columns that are parameters are materialized
+  for (const auto& parameter : expression.parameters) {
+    _ensure_column_materialization(parameter.second);
+  }
+
+  std::vector<std::shared_ptr<const Table>> results(_output_row_count);
+
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < _output_row_count; ++chunk_offset) {
+    results[chunk_offset] = _evaluate_select_expression_for_row(expression, chunk_offset);
+  }
+
+  return results;
+}
+
+std::shared_ptr<const Table> ExpressionEvaluator::_evaluate_select_expression_for_row(
+    const PQPSelectExpression& expression, const ChunkOffset chunk_offset) {
+  Assert(expression.parameters.empty() || _chunk,
+         "Sub-SELECT references external Columns but Expression doesn't operate on a Table/Chunk");
+
+  std::unordered_map<ParameterID, AllTypeVariant> parameters;
+
+  for (auto parameter_idx = size_t{0}; parameter_idx < expression.parameters.size(); ++parameter_idx) {
+    const auto& parameter_id_column_id = expression.parameters[parameter_idx];
+    const auto parameter_id = parameter_id_column_id.first;
+    const auto column_id = parameter_id_column_id.second;
```

Comment:
structured binding?

Reply:
Don't like them because they encourage pythong style programming. Also, my compiler doesn't support them yet. Leaving this one out, sorry ;)